credentials/tls_ext_test.go

```diff
@@ -79,43 +79,80 @@ func (s) TestTLS_MinVersion12(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
 
-	// Create server creds without a minimum version.
-	serverCreds := credentials.NewTLS(&tls.Config{
+	serverTLS := &tls.Config{
 		// MinVersion should be set to 1.2 by gRPC by default.
 		Certificates: []tls.Certificate{serverCert},
-	})
-	ss := stubserver.StubServer{
-		EmptyCallF: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
-			return &testpb.Empty{}, nil
+	}
+
+	testCases := []struct {
+		name      string
+		serverTLS func() *tls.Config
+	}{
+		{
+			name:      "base_case",
+			serverTLS: func() *tls.Config { return serverTLS },
```

Comment:
Structure question - why define a `func() *tls.Config` instead of just having the field be `tls.Config`? It seems like it's just adding a layer of complication but doesn't add anything?

Same question throughout

Reply:
Earlier I had defined a `tls.Config` that returns `nil, nil` from `GetConfigForClient` as a variable above the test case slice:
```go
noOpGetCfgForClient := serverTLS.Clone()
noOpGetCfgForClient.GetConfigForClient = func(*tls.ClientHelloInfo) (*tls.Config, error) {
	return nil, nil
}

testCases := []struct {
// Omitted.
}
```

In comment https://github.com/grpc/grpc-go/pull/7754#discussion_r1805589806, Easwar suggested using a function to generate the config can get rid of the variable in the outer scope, so I made the change https://github.com/grpc/grpc-go/pull/7754/commits/0b1146d634efba63fe6d815ea3084dd351b37fff.

```go
testCases := []struct {
// Omitted.
			serverTLS: func() *tls.Config {
				config := serverTLS.Clone()
				config.GetConfigForClient = func(*tls.ClientHelloInfo) (*tls.Config, error) {
					return nil, nil
				}
				return config
			},
// Omitted.

Reply:
What I meant was just fully defining the configs in the test case definition, that way when reading the test code there's no indirection to other parts or higher scoped variables, cloning, or anything. You just make the config you want for that test. Especially since there's only one line of boilerplate `Certificates: []tls.Certificates(serverCert)` per config

i.e.

```
testCases := []struct {
		name      string
		serverTLS tls.Config
	}{
		{
			name:      "base_case",
			serverTLS: tls.Config {
                                 Certificates: []tls.Certificate{serverCert},
                         }
                 },
                 {
                        name: "fallback_to_base"
                        serverTLS: tls.Config {
                                  Certificates: []tls.Certificate{serverCert},
                                  GetConfigForClient = func(*tls.ClientHelloInfo) (*tls.Config, error) {
					return nil, nil
				}
                                 // any other options you want to set, etc
			}
                  },
                  // more cases
        }
```

Hopefully that is clear - I'm good for you to merge either way, pick whichever way you think best

Reply:
I agree, inlining the configs makes the test easier to read. Changes made.